### PR TITLE
Raise `StepExecutionError` on non-zero `run_shell_script` return code during `prefect deploy`

### PR DESCRIPTION
**Pull request by serinamarie**
This PR focuses on handling errors more effectively during the execution of `run_shell_script` in deployment steps. Instead of silently allowing deployment steps to continue during `prefect deploy`, we raise on error here. 

### Checklist

- [x] This pull request references any related issue by including "closes `<link to issue>`"
	- If no issue exists and your change is not a small fix, please [create an issue](https://github.com/PrefectHQ/prefect/issues/new/choose) first.
- [x] This pull request includes tests or only affects documentation.
- [x] This pull request includes a label categorizing the change e.g. `maintenance`, `fix`, `feature`, `enhancement`, `docs`.
  

For documentation changes:

- [ ] This pull request includes redirect settings in `netlify.toml` for files that are removed or renamed

### CODE REVIEW

**Comment by serinamarie** [comment]
Curious to see all the tests that fail 😅

File: src/prefect/deployments/steps/utility.py [hunk 1]
```diff
@@ -188,6 +188,12 @@ async def run_shell_script(
 
             await process.wait()
 
+            if process.returncode != 0:
+                raise RuntimeError(
+                    f"run_shell_script failed with error code {process.returncode}:"
```
Comment (zzstoatzz) [inline src/prefect/deployments/steps/utility.py:193]: nit
```suggestion
                    f"`run_shell_script` failed with error code {process.returncode}:"
```

File: tests/cli/test_deploy.py [hunk 2]
```diff
@@ -2397,6 +2398,39 @@ async def test_deploy_does_not_prompt_build_docker_image_when_empty_build_action
             "An important name/test-name"
         )
 
+    async def test_deploy_with_bad_run_shell_script_raises(
+        self, project_dir, work_pool
+    ):
+        """
+        Regression test for a bug where deployment steps woould continue even when
```
Comment (zzstoatzz) [inline tests/cli/test_deploy.py:2405]:
```suggestion
        Regression test for a bug where deployment steps would continue even when
```

File: tests/cli/test_deploy.py [hunk 3]
```diff
@@ -2397,6 +2398,39 @@ async def test_deploy_does_not_prompt_build_docker_image_when_empty_build_action
...
+        """
+        prefect_file = Path("prefect.yaml")
+        with prefect_file.open(mode="r") as f:
+            config = yaml.safe_load(f)
+
+        config["build"] = [
+            {
+                "prefect.deployments.steps.run_shell_script": {
+                    "id": "test",
+                    "script": "cat nothing",
```
Comment (zzstoatzz) [inline tests/cli/test_deploy.py:2416]: 🌫️ 🐈

**Review by zzstoatzz — APPROVED** [review]
tiny nits but LGTM!

File: tests/cli/test_deploy.py [hunk 4]
```diff
@@ -6006,8 +6040,8 @@ async def test_other_build_step_exists_prompts_build_custom_docker_image(
         prefect_config["build"] = [
             {
                 "prefect.deployments.steps.run_shell_script": {
-                    "id": "get-commit-hash",
-                    "script": "git rev-parse --short HEAD",
+                    "id": "sample-bash-cmd",
+                    "script": "echo 'Hello, World!'",
```
Comment (serinamarie) [inline tests/cli/test_deploy.py:6044]: needed to adjust this as it would fail on a 128 otherwise

**Pull request merged**